### PR TITLE
[Tabs] Switch to manual layout

### PR DESCRIPTION
**Pull request by romoore**
Adding the selection indicator was creating a lot of headache using autolayout. This PR switches to manual layout instead.

Part of #7744

### CODE REVIEW

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 1]
```diff
@@ -443,73 +418,91 @@ - (void)layoutSubviews {
...
+  [super willMoveToSuperview:newSuperview];
+  self.initialScrollDone = NO;
 }
 
 - (CGSize)intrinsicContentSize {
+  //  if (self.isJustifiedLayoutStyle) {
+  return [self intrinsicContentSizeForJustifiedLayout];
+  //  } else {
+  //    return [self intrinsicContentSizeForScrollableLayout];
+  //  }
```
Comment (mikefan1991) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Remove the comments if not needed.

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 2]
```diff
@@ -443,73 +418,91 @@ - (void)layoutSubviews {
   }
 }
 
-- (void)willMoveToSuperview:(UIView *)newSuperview {
-  [super willMoveToSuperview:newSuperview];
-  self.initialScrollDone = NO;
+- (BOOL)isJustifiedLayoutStyle {
```
Comment (mikefan1991) [inline components/Tabs/src/TabBarView/MDCTabBarView.m:433]: Since the getter is called several times. I think it is better to override the setter and use the self.isJustifiedLayoutStyle as a BOOL flag. But this can happen later.

**Review by mikefan1991 — APPROVED** [review]

**Review by leonmz — COMMENTED** [review]
Fantastic work! Let me play with this for a bit  :)

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 3]
```diff
@@ -443,73 +418,91 @@ - (void)layoutSubviews {
...
-}
 
-- (void)updateConstraints {
-  if (self.containerViewConstraintsActive) {
-    [super updateConstraints];
-    return;
+  BOOL isRTL =
+      [self mdf_effectiveUserInterfaceLayoutDirection] == UIUserInterfaceLayoutDirectionRightToLeft;
+
+  CGFloat itemViewWidth = CGRectGetWidth(availableBounds) / self.itemViews.count;
```
Comment (leonmz) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: If the available width is 1500 and the tab bar has 3 items then the item will being stretched to 500.
Is this what we expected? Or we need to switch to scrollable view in this case..

**Review by leonmz — APPROVED** [review]

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 2 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m:433]: I know it's a bit more computationally heavy, but for now I do prefer that it's a calculated value. That ensures we don't forget to update it and it always reflects the current state of the layout.  I agree that later we can make some internal optimizations.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 1 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Done, thanks!

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 3 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Unless you think it needs to be changed for this version, I'd like to leave the behavior as-is for now.  This logic is simpler and less likely to be wrong.  I think we can improve it in the future, especially since in that case I would expect clients to chose "Clustered Fixed" layout as their preferred layout style.

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 4]
```diff
@@ -420,100 +406,109 @@ - (void)observeValueForKeyPath:(NSString *)keyPath
...
     self.initialScrollDone = YES;
     [self scrollUntilSelectedItemIsVisibleWithoutAnimation];
   }
 }
 
-- (void)willMoveToSuperview:(UIView *)newSuperview {
-  [super willMoveToSuperview:newSuperview];
-  self.initialScrollDone = NO;
+- (BOOL)isJustifiedLayoutStyle {
+  CGRect availableBounds = self.bounds;
```
Comment (codeman7) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Tip: Refactor this out to a shared method

```objc
- (CGRect)availableBounds {
  CGRect availableBounds = self.bounds;
  if (@available(iOS 11.0, *)) {
    availableBounds = UIEdgeInsetsInsetRect(availableBounds, self.adjustedContentInset);
  }
  return availableBounds;
}
```

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 5]
```diff
@@ -420,100 +406,109 @@ - (void)observeValueForKeyPath:(NSString *)keyPath
...
-    self.scrollableLayoutConstraints = @[];
-    [self.leadingAnchor constraintEqualToAnchor:self.containerView.leadingAnchor].active = YES;
-    [self.topAnchor constraintEqualToAnchor:self.containerView.topAnchor].active = YES;
-    [self.trailingAnchor constraintEqualToAnchor:self.containerView.trailingAnchor].active = YES;
-    [self.bottomAnchor constraintEqualToAnchor:self.containerView.bottomAnchor].active = YES;
+    availableBounds = UIEdgeInsetsInsetRect(availableBounds, self.adjustedContentInset);
   }
-  self.containerViewConstraintsActive = YES;
+  BOOL isRTL =
+      [self mdf_effectiveUserInterfaceLayoutDirection] == UIUserInterfaceLayoutDirectionRightToLeft;
```
Comment (codeman7) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Tip: Please use dot notation.

**Review by codeman7 — APPROVED** [review]

File: components/Tabs/examples/MDCTabBarViewTypicalExampleViewController.m [hunk 6]
```diff
@@ -74,6 +74,9 @@ - (void)viewDidLoad {
   [super viewDidLoad];
   self.title = kExampleTitle;
 
+  UIScrollView *dummyScrollView = [[UIScrollView alloc] init];
+  [self.view addSubview:dummyScrollView];
```
Comment (jverkoey) [inline components/Tabs/examples/MDCTabBarViewTypicalExampleViewController.m]: Please document that this is being added so that the app bar does not inject content insets into the tab bar. Please also pull this out to a separate PR.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 4 above)
Comment (jverkoey) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: +1

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 3 above)
Comment (jverkoey) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Can this code be reached if the number of items is zero?

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 7]
```diff
@@ -420,100 +406,109 @@ - (void)observeValueForKeyPath:(NSString *)keyPath
...
+    CGSize intrinsicContentSize = view.intrinsicContentSize;
+    view.frame =
+        CGRectMake(itemViewOriginX, itemViewOriginY, intrinsicContentSize.width, itemViewHeight);
+    itemViewOriginX += intrinsicContentSize.width;
+  }
+}
+
+- (void)willMoveToSuperview:(UIView *)newSuperview {
+  [super willMoveToSuperview:newSuperview];
+  self.initialScrollDone = NO;
```
Comment (jverkoey) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Consider changing this to “needsScrollToSelectedItem”

**Review by jverkoey — APPROVED** [review]

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 7 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Done

File: components/Tabs/examples/MDCTabBarViewTypicalExampleViewController.m (hunk 6 above)
Comment (romoore) [reply components/Tabs/examples/MDCTabBarViewTypicalExampleViewController.m]: Doing in #7772

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 5 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Done.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 4 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Done.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 3 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: No longer. Done.
